### DISCUSSION TITLE
WebUI Improvement suggestions

### DISCUSSION THREAD

**Discussion by pbiering** [Kozea/Radicale#1416] [Ideas]
@MatthewHana: would it be possible to enrich the new WebUI by:

- [ ] "create collection": input box for optional directory name, replacing in this case the `random_uuid` used in `CreateEditCollectionScene`
- [ ] "list collections": display amount of entries per collection
- [ ] "delete collection": display amount of entries in particular collection and request explicit typing "CONFIRM" into an additional input box
- [ ] "edit collection": is there any support in the protocol to adjust the "C:supported-calendar-component-set" via client after a collection was created? If so, can this also be implemented?

**Comment by MatthewHana** [comment]
> @MatthewHana: would it be possible to enrich the new WebUI by:
> 
> - [x] "create collection": input box for optional directory name, replacing in this case the `random_uuid` used in `CreateEditCollectionScene`

Yes, I had actually considered how to implement this in my previous UI changes but didn't know if people would want it. Can easily be implemented.

> - [x] "list collections": display amount of entries per collection

I'll have a look into the DAV specs to see how I can efficiently implement this.

> - [x] "delete collection": display amount of entries in particular collection and request explicit typing "CONFIRM" into an additional input box

I can implement this easily.

> - [x] "edit collection": is there any support in the protocol to adjust the "C:supported-calendar-component-set" via client after a collection was created? If so, can this also be implemented?

I will have a look into this and find out.

**Reply by MatthewHana** [reply]
I have implemented:
- Typing a HREF name for a new collection
- Requiring the user to ttype 'DELETE' into a textbox on the delete confirmation dialog
- Changing component sets for Calendars
- Getting the number of entries in a collection (needs PR #1424)

What I need input on before I have a PR ready is **where should the number of entries for a collection be displayed**? I considered the following:
- In parenthesis after the collection type
- At the end of the description
- On a new line below the collection type

**Screenshots**
<img width="352" alt="new-radicale-calsets" src="https://github.com/Kozea/Radicale/assets/1935851/1c1a20dd-493b-4789-aefa-ddfa4553e141">
<img width="337" alt="new-radicale-del" src="https://github.com/Kozea/Radicale/assets/1935851/72b65050-e790-4c5e-be02-3766f88c8bb4">
<img width="358" alt="new-radicale-href" src="https://github.com/Kozea/Radicale/assets/1935851/5c56de62-1688-4a0d-bd27-787627395aee">

**Comment by MatthewHana** [comment]
**Solution for list collections**
I've had a look into displaying the number of entries within a collection. There does not appear to be a DAV property to query the number of items in a collection, but there is a DAV property to obtain the character length of the total collection. 

To implement this I created a custom Radicale DAV property called ```getcontentcount``` (named to be consistent with the standard DAV get* properties) to get the total number of items in a collection.

Pro of this approach: It gets the job done
Con of this approach: Deviation from the DAV standards for both the Radicale server and the Radicale Web UI.
Things to consider: Radicale already has a custom property to get the display name for a collection for the Web UI.

If there are no objections to this approach then I will raise a separate PR to add this property.

<img width="454" alt="getcontentcount" src="https://github.com/Kozea/Radicale/assets/1935851/a21f7370-f0ab-4180-9c90-b62286aa39d7">

**Reply by pbiering** [reply]
as there is already one custom property, I would see no issue adding a 2nd one. But potentially it would be a good idea to document them in the Wiki on a new page...potentially other find this useful and take over and create new draft for upcoming RFC updates out of it...

**Reply by MatthewHana** [reply]
Hi @pbiering, I've made a PR to add the custom prop. See PR #1424

**Comment by pbiering** [comment]
> What I need input on before I have a PR ready is **where should the number of entries for a collection be displayed**? I considered the following:
>     * On a new line below the collection type

I would prefer this between the collection type and the URL like: `x items`

Also is there a capability available of some kind of mouse-over/click support that "copy" hint will appear?

**Reply by MatthewHana** [reply]
Can you explain further what you mean by a copy hint?

**Reply by MatthewHana** [reply]
Something like this?
<img width="221" alt="contentcount" src="https://github.com/Kozea/Radicale/assets/1935851/c432683f-a465-4c5a-875c-7ae593639772">

**Reply by pbiering** [reply]
that  looks nice!

**Comment by MatthewHana** [comment]
New PR #1432 implementing all improvements.

**Comment by pbiering** [comment]
Thank you very much! One last idea:

> character length of the total collection

can this be used to display a size of collection nearby the amount of items, best in human readable (rounded-up) format like

```
92 items in collection (220 kByte)
```

**Reply by MatthewHana** [reply]
See PR #1433

**Comment by pbiering** [comment]
Thank you, one last found during testing: what about a "refresh button", because page reload results in login page again.

**Reply by MatthewHana** [reply]
I will get to this in a couple of days.

**Reply by MatthewHana** [reply]
Implemented. Will make PR in a couple of hours.

**Comment by pbiering** [comment]
And should I be able to adjust or delete an uploaded calendar? If yes, then something is strange now, no "mouse over"...

Tested with: `http://i.cal.to/ical/65/bayern/feiertage/499a5ad5.07785850-d8e6092a.ics`

**Reply by MatthewHana** [reply]
Are the icons no longer appearing on mouse hover?

**Reply by pbiering** [reply]
No, but now reproduced, it's depending on the length of "Title", if wrapping into 2 lines but still large font is used, the "mouse over" is gone.
![Bildschirmfoto_2024-03-13_07-25-41](https://github.com/Kozea/Radicale/assets/8064668/15a7c4a7-2965-4eb7-a1ad-9282cb191dc3)

**Reply by MatthewHana** [reply]
I see. I'll tweak the CSS to make it more responsive in these cases. I'll push a change out tomorrow.

**Comment by pbiering** [comment]
- [ ] potentially "in collection" can be removed to avoid that even this line is suddenly wrapping
- [ ] would it be easy to implement optional custom "href" also for upload?

**Reply by MatthewHana** [reply]
I can add HREF textbox for uploading. I'll include it in the next batch of UI improvements I'll push out in a few days.

**Reply by MatthewHana** [reply]
Implemented. Surprisingly more annoying than I originally anticipated. Limitation is that HREF can only be set when uploading 1 file only, otherwise the HREF box will be hidden and random name will be applied. Will make PR in a couple of hours.

**Reply by MatthewHana** [reply]
New PR #1456 to implement new features + fixes.

**Comment by pbiering** [comment]
@MatthewHana : can you please contribute details of the "radicale" related known extensions?

https://github.com/Kozea/Radicale/wiki/Radicale-Custom-DAV-protocol-extensions

Thank you very much!

**Reply by MatthewHana** [reply]
Done!

**Reply by pbiering** [reply]
Thank you very much!

**Comment by MatthewHana** [comment]
What I would like to do next is a holistic refactoring of the web client JavaScript to make it easier to add functionality into the future. I also want to split it into multiple JS to make it easier to make and track changes to the code. Currently UI, DAV, and Helper functions are all in the same file- I would like to split them up and eventually implement JS testing. Will post progress here.

**Comment by pbiering** [comment]
Refactoring done by @maxberger
